lib/spack/spack/schema/modules.py

```diff
@@ -9,110 +9,110 @@
    :lines: 13-
 """
 
+#: Matches a spec or a multi-valued variant but not another
+#: valid keyword.
+#:
+#: THIS NEEDS TO BE UPDATED FOR EVERY NEW KEYWORD THAT
+#: IS ADDED IMMEDIATELY BELOW THE MODULE TYPE ATTRIBUTE
+spec_regex = r'(?!hierarchy|verbose|hash_length|whitelist|' \
+             r'blacklist|naming_scheme|core_compilers|all)(^\w[\w-]*)'
+
+#: Matches an anonymous spec, i.e. a spec without a root name
+anonymous_spec_regex = r'^[\^@%+~]'
 
 #: Definitions for parts of module schema
-definitions = {
-    'array_of_strings': {
-        'type': 'array',
-        'default': [],
-        'items': {
-            'type': 'string'
-        }
-    },
-    'dictionary_of_strings': {
-        'type': 'object',
-        'patternProperties': {
-            r'\w[\w-]*': {  # key
-                'type': 'string'
-            }
-        }
-    },
-    'dependency_selection': {
-        'type': 'string',
-        'enum': ['none', 'direct', 'all']
-    },
-    'module_file_configuration': {
-        'type': 'object',
-        'default': {},
-        'additionalProperties': False,
-        'properties': {
-            'filter': {
-                'type': 'object',
-                'default': {},
-                'additionalProperties': False,
-                'properties': {
-                    'environment_blacklist': {
-                        'type': 'array',
-                        'default': [],
-                        'items': {
-                            'type': 'string'
-                        }
+array_of_strings = {
+    'type': 'array', 'default': [], 'items': {'type': 'string'}
+}
+
+dictionary_of_strings = {
+    'type': 'object', 'patternProperties': {r'\w[\w-]*': {'type': 'string'}}
+}
```

Comment:
@tgamblin I believe all comments have been addressed. Doing this refactor I saw a couple of minor things that you might want to do or not (either here or in another PR):

1. `array_of_strings` and `dictionary_of_strings` might be generic enough to be moved into `spack/schema/__init__.py` and be used in schema other than modules
2. `packages.yaml` could probably make use of `validate_spec` too

Let me know what do you think about these.